[SPARK-23690][ML] Add handleinvalid to VectorAssembler

yogeshg opened the pull request:
## What changes were proposed in this pull request?

Introduce `handleInvalid` parameter in `VectorAssembler` that can take in `"keep", "skip", "error"` options. "error" throws an error on seeing a row containing a `null`, "skip" filters out all such rows, and "keep" adds relevant number of NaN. "keep" figures out an example to find out what this number of NaN s should be added and throws an error when no such number could be found.

## How was this patch tested?

Unit tests are added to check the behavior of `assemble` on specific rows and the transformer is called on `DataFrame`s of different configurations to test different corner cases.

Commit:
add NaN for null column

Commit:
get lengths with a map

MrBago committed:
Better error for streaming dataframes, ensure non-null Vectors in first.

Commit:
wip

Commit:
wip

Commit:
wip

MrBago committed:
Merge branch 'rformula_handleinvalid' into vectorAssemblerStuff

MrBago committed:
Merge fixes.

yogeshg committed:
Merge pull request #2 from MrBago/vectorAssemblerStuff

Vector assembler stuff

Commit:
fix issues with this implementation

Commit:
fix bugs; add tests

Commit:
clean

AmplabJenkins commented:
Can one of the admins verify this patch?

SparkQA commented:
**[Test build #88249 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88249/testReport)** for PR 20829 at commit [`c0c0e3d`](https://github.com/apache/spark/commit/c0c0e3df92dc787d9ce9ca632353d86c6457420d).

SparkQA commented:
**[Test build #88249 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88249/testReport)** for PR 20829 at commit [`c0c0e3d`](https://github.com/apache/spark/commit/c0c0e3df92dc787d9ce9ca632353d86c6457420d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88249/
Test FAILed.

WeichenXu123 commented on mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala:
```diff
@@ -234,7 +234,7 @@ class StringIndexerModel (
     val metadata = NominalAttribute.defaultAttr
       .withName($(outputCol)).withValues(filteredLabels).toMetadata()
     // If we are skipping invalid records, filter them out.
-    val (filteredDataset, keepInvalid) = getHandleInvalid match {
```
Why need change this line ?

yogeshg replied:
Thanks for picking this out! I changed this because I was matching on `$(handleInvalid)` in VectorAssembler and that seems to be the recommended way of doing this. Should I include this in the current PR and add a note or open a separate PR?

WeichenXu123 replied:
ok. it doesn't matter no need separate PR I think. just a minor change.

jkbradley replied:
For the record, in general, I would not bother making changes like this.  The one exception I do make is IntelliJ style complaints since those can be annoying for developers.

Commit:
edge case of all null column to be skipped handled iff no elements found

SparkQA commented:
**[Test build #88277 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88277/testReport)** for PR 20829 at commit [`bf2f5b3`](https://github.com/apache/spark/commit/bf2f5b39b3056b2301afd1e19582fea0e3700f3a).

SparkQA commented:
**[Test build #88277 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88277/testReport)** for PR 20829 at commit [`bf2f5b3`](https://github.com/apache/spark/commit/bf2f5b39b3056b2301afd1e19582fea0e3700f3a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88277/
Test FAILed.

Commit:
update logic for missing metadata

yogeshg commented:
I fixed code paths that failed tests, waiting for @SparkQA . Offline talk with @MrBago suggests that we can perhaps decrease the number of maps in `transform` method. Looking into that.

SparkQA commented:
**[Test build #88278 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88278/testReport)** for PR 20829 at commit [`5ce7671`](https://github.com/apache/spark/commit/5ce7671267a4dd240b80b6230e9e3729c96de7e6).

SparkQA commented:
**[Test build #88278 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88278/testReport)** for PR 20829 at commit [`5ce7671`](https://github.com/apache/spark/commit/5ce7671267a4dd240b80b6230e9e3729c96de7e6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88278/
Test PASSed.

Commit:
rafactor logic to simplify code paths

SparkQA commented:
**[Test build #88286 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88286/testReport)** for PR 20829 at commit [`482225f`](https://github.com/apache/spark/commit/482225fb1ffeccc475972e42246b5f8681d7fee5).

yogeshg commented:
test this please

SparkQA commented:
**[Test build #88287 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88287/testReport)** for PR 20829 at commit [`482225f`](https://github.com/apache/spark/commit/482225fb1ffeccc475972e42246b5f8681d7fee5).

SparkQA commented:
**[Test build #88286 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88286/testReport)** for PR 20829 at commit [`482225f`](https://github.com/apache/spark/commit/482225fb1ffeccc475972e42246b5f8681d7fee5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88286/
Test PASSed.

SparkQA commented:
**[Test build #88287 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88287/testReport)** for PR 20829 at commit [`482225f`](https://github.com/apache/spark/commit/482225fb1ffeccc475972e42246b5f8681d7fee5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88287/
Test PASSed.

WeichenXu123 commented on mllib/src/main/scala/org/apache/spark/ml/feature/VectorAssembler.scala:
```diff
@@ -85,18 +120,34 @@ class VectorAssembler @Since("1.4.0") (@Since("1.4.0") override val uid: String)
+    val lengths = featureAttributesMap.map(a => a.length)
+    val metadata = new AttributeGroup($(outputCol), featureAttributes.toArray).toMetadata()
+    val (filteredDataset, keepInvalid) = $(handleInvalid) match {
+      case StringIndexer.SKIP_INVALID => (dataset.na.drop("any", $(inputCols)), false)
```
you can directly use `dataset.na.drop($(inputCols))`

yogeshg replied:
Ah, good point! Although I do think that keeping "any" might make it easier to read, but that may not necessarily hold for experienced people :P

WeichenXu123 commented on mllib/src/main/scala/org/apache/spark/ml/feature/VectorAssembler.scala:
```diff
@@ -49,32 +51,65 @@ class VectorAssembler @Since("1.4.0") (@Since("1.4.0") override val uid: String)
+           """.stripMargin.replaceAll("\n", " "))
+          }
+          if (isMissingNumAttrs) {
+            val column = dataset.select(c).na.drop()
```
* The var name `column` isn't good. `colDataset` is better.

* An optional optimization is one-pass scanning the dataset and count non-null rows for each "missing num attrs" columns.

hootoconnor commented:
What A Mess !

On Mar 14, 2018 8:17 PM, "UCB AMPLab" <notifications@github.com> wrote:

> Merged build finished. Test FAILed.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/spark/pull/20829#issuecomment-373247488>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/Ad_gynMT5mf8PhL_CV3DvvUzLzcUfzKvks5ted1VgaJpZM4Srd7a>
> .
>

Commit:
update logic as discussed with Bago

Commit:
fix test cases, remove some conditionals, move to static methods in companion object

SparkQA commented:
**[Test build #88390 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88390/testReport)** for PR 20829 at commit [`2b1fd4e`](https://github.com/apache/spark/commit/2b1fd4e36fffefb1e1679221353426cb05320104).

Commit:
use the correct statics, update since

SparkQA commented:
**[Test build #88392 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88392/testReport)** for PR 20829 at commit [`9624061`](https://github.com/apache/spark/commit/9624061920bfe0a2cf06af054d9988c95cd0b322).

SparkQA commented:
**[Test build #88390 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88390/testReport)** for PR 20829 at commit [`2b1fd4e`](https://github.com/apache/spark/commit/2b1fd4e36fffefb1e1679221353426cb05320104).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88390/
Test PASSed.

SparkQA commented:
**[Test build #88392 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88392/testReport)** for PR 20829 at commit [`9624061`](https://github.com/apache/spark/commit/9624061920bfe0a2cf06af054d9988c95cd0b322).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88392/
Test PASSed.

Commit:
collect instead of cache

SparkQA commented:
**[Test build #88395 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88395/testReport)** for PR 20829 at commit [`ab91545`](https://github.com/apache/spark/commit/ab91545bebc6e1d0c5c3cb7c15156d546ad48f81).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88395/
Test PASSed.

jkbradley commented:
@hootoconnor Please refrain from making non-constructive comments.  If you did not intend to leave the comment here, please remove it.  Thanks.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/feature/VectorAssembler.scala:
```diff
@@ -136,34 +172,88 @@ class VectorAssembler @Since("1.4.0") (@Since("1.4.0") override val uid: String)
   override def load(path: String): VectorAssembler = super.load(path)
 
-  private[feature] def assemble(vv: Any*): Vector = {
+  private[feature] def assemble(lengths: Seq[Int], keepInvalid: Boolean)(vv: Any*): Vector = {
```
nit: Use Array[Int] for faster access

jkbradley replied:
Also, I'd add doc explaining requirements, especially that this assumes that lengths and vv have the same length.

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/feature/VectorAssemblerSuite.scala:
```diff
@@ -37,24 +37,26 @@ class VectorAssemblerSuite
   test("assemble should compress vectors") {
     import org.apache.spark.ml.feature.VectorAssembler.assemble
-    val v1 = assemble(0.0, 0.0, 0.0, Vectors.dense(4.0))
+    val v1 = assemble(Seq(1, 1, 1, 4), true)(0.0, 0.0, 0.0, Vectors.dense(4.0))
```
We probably want this to fail, right?  It expects a Vector of length 4 but is given a Vector of length 1.

yogeshg replied:
that's a typo, Thanks for pointing it out! that number is not used in case we do not have nulls, which is why the test passes

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/feature/VectorAssembler.scala:
```diff
@@ -49,32 +53,57 @@ class VectorAssembler @Since("1.4.0") (@Since("1.4.0") override val uid: String)
+  def setHandleInvalid(value: String): this.type = set(handleInvalid, value)
+
+  /**
+   * Param for how to handle invalid data (NULL values). Options are 'skip' (filter out rows with
```
It would be good to expand this doc to explain the behavior: how various types of invalid values are treated (null, NaN, incorrect Vector length) and how computationally expensive different options can be.

yogeshg replied:
Behavior of options already included, explanation of column length included here, run time information included in the VectorAssembler class's documentation. Thanks for the suggestion, this is super important!

yogeshg replied:
also, we just deal with nulls here. NaNs and incorrect length vectors are transmitted transparently. Do we need to test for those?

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/feature/VectorAssemblerSuite.scala:
```diff
@@ -147,4 +149,72 @@ class VectorAssemblerSuite
       .count() == 1)
   }
+
+  test("assemble should keep nulls") {
```
make more explicit: + " when keepInvalid = true"

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/feature/VectorAssemblerSuite.scala:
```diff
@@ -147,4 +149,72 @@ class VectorAssemblerSuite
+    assert(assemble(Seq(2), true)(null) === Vectors.dense(Double.NaN, Double.NaN))
+  }
+
+  test("assemble should throw errors") {
```
similarly: make more explicit: + " when keepInvalid = false"

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/feature/VectorAssemblerSuite.scala:
```diff
@@ -147,4 +149,72 @@ class VectorAssemblerSuite
+  test("assemble should throw errors") {
+    import org.apache.spark.ml.feature.VectorAssembler.assemble
+    intercept[SparkException](assemble(Seq(1, 1), false)(1.0, null) ===
+      Vectors.dense(1.0, Double.NaN))
```
No need to compare with anything; just call assemble()

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/feature/VectorAssemblerSuite.scala:
```diff
@@ -147,4 +149,72 @@ class VectorAssemblerSuite
+  }
+
+  test("get lengths function") {
+    val df = Seq[(Long, Long, java.lang.Double, Vector, String, Vector, Long)](
```
Why java.lang.Double instead of Double?

yogeshg replied:
to allow nulls in the column :)

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/feature/VectorAssembler.scala:
```diff
@@ -136,34 +172,88 @@ class VectorAssembler @Since("1.4.0") (@Since("1.4.0") override val uid: String)
+    Array(SKIP_INVALID, ERROR_INVALID, KEEP_INVALID)
+
+
+  private[feature] def getLengthsFromFirst(dataset: Dataset[_],
```
scala style: For multiline class and method headers, put the first argument on the next line, with +4 space indentation

jkbradley replied:
Also, it's unclear what this method does until I read the code.  Possibly make the method name more explicit ("getVectorLengthsFromFirstRow"), and definitely document that "columns" must all be of Vector type.

yogeshg replied:
updated

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/feature/VectorAssembler.scala:
```diff
@@ -136,34 +172,88 @@ class VectorAssembler @Since("1.4.0") (@Since("1.4.0") override val uid: String)
+  private[feature] def getLengthsFromFirst(dataset: Dataset[_],
+                                           columns: Seq[String]): Map[String, Int] = {
+    try {
+      val first_row = dataset.toDF.select(columns.map(col): _*).first
```
Fix (new) IntelliJ style warnings: "toDF" -> "toDF()"

jkbradley replied:
first -> first()

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/feature/VectorAssembler.scala:
```diff
@@ -136,34 +172,88 @@ class VectorAssembler @Since("1.4.0") (@Since("1.4.0") override val uid: String)
+      case (true, VectorAssembler.ERROR_INVALID) =>
+        getLengthsFromFirst(dataset, missing_columns)
+      case (true, VectorAssembler.SKIP_INVALID) =>
+        getLengthsFromFirst(dataset.na.drop, missing_columns)
```
This will drop Rows with NA values in extraneous columns.  I.e., even if the VectorAssembler is only assembling columns A and B, if there is a NAN in column C, this will drop that row.  Pass the list of columns you care about to the drop() method.

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/feature/VectorAssemblerSuite.scala:
```diff
@@ -147,4 +149,72 @@ class VectorAssemblerSuite
+      (3, 3, null, Vectors.dense(1.0, 2.0), "a", Vectors.sparse(2, Array(1), Array(3.0)), 8L),
+      (4, 4, null, null, "a", Vectors.sparse(2, Array(1), Array(3.0)), 9L)
+    ).toDF("id1", "id2", "x", "y", "name", "z", "n")
+    assert(VectorAssembler.getLengthsFromFirst(df, Seq("y")).exists(_ == "y" -> 2))
```
I think you can do a direct comparison:
```
    assert(VectorAssembler.getLengthsFromFirst(df, Seq("y")) === Map("y" -> 2))
```

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/feature/VectorAssemblerSuite.scala:
```diff
@@ -147,4 +149,72 @@ class VectorAssemblerSuite
+      Vectors.dense(Double.NaN, Double.NaN))
+  }
+
+  test("get lengths function") {
```
This is great that you're testing this carefully, but I recommend we make sure to pass better exceptions to users.  E.g., they won't know what to do with a NullPointerException, so we could instead tell them something like: "Column x in the first row of the dataset has a null entry, but VectorAssembler expected a non-null entry.  This can be fixed by explicitly specifying the expected size using VectorSizeHint."

yogeshg replied:
Thanks! We do throw some descriptive error [here](https://github.com/apache/spark/pull/20829/files#diff-9c84e1d27f25714e256cb482069359cfR193), added more description to it and made assertions in test on those messages.

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/feature/VectorAssemblerSuite.scala:
```diff
@@ -147,4 +149,72 @@ class VectorAssemblerSuite
+  }
+
+  test("Handle Invalid should behave properly") {
+    val df = Seq[(Long, Long, java.lang.Double, Vector, String, Vector, Long)](
```
Since this is shared across multiple tests, just make it a shared value.  See e.g. https://github.com/apache/spark/blob/master/mllib/src/test/scala/org/apache/spark/ml/classification/LogisticRegressionSuite.scala#L55

jkbradley replied:
Also, if there are "trash" columns not used by VectorAssembler, maybe name them as such and add a few null values in them for better testing.

yogeshg replied:
thanks, good idea! this helped me in catching the `drop.na()` bug that might drop everything

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/feature/VectorAssemblerSuite.scala:
```diff
@@ -147,4 +149,72 @@ class VectorAssemblerSuite
+
+    // behavior when first row has information
+    assert(assembler.setHandleInvalid("skip").transform(df).count() == 1)
+    intercept[RuntimeException](assembler.setHandleInvalid("keep").transform(df).collect())
```
Should this fail?  I thought it should pad with NaNs.

yogeshg replied:
it fails because vector size hint is not given, adding a section with VectorSizeHInts

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/feature/VectorAssemblerSuite.scala:
```diff
@@ -147,4 +149,72 @@ class VectorAssemblerSuite
+    intercept[RuntimeException](assembler.setHandleInvalid("keep").transform(df).collect())
+    intercept[SparkException](assembler.setHandleInvalid("error").transform(df).collect())
+
+    // numeric column is all null
```
Did you want to test:
* extraction of metadata from the first row (which is what this is testing, I believe), or
* transformation on an all-null column (which this never reaches)?

yogeshg replied:
was testing extraction of metadata for numeric column (is always 1). Not relevant in new framework.

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/feature/VectorAssemblerSuite.scala:
```diff
@@ -147,4 +149,72 @@ class VectorAssemblerSuite
+    intercept[RuntimeException](
+      assembler.setHandleInvalid("keep").transform(df.filter("id1==3")).count() == 1)
+
+    // vector column is all null
```
ditto

jkbradley reviewed (commented):
Thanks for the PR!  I made a pass, mainly looking at tests.

Commit:
update test logic; bugfix for na.drop; add more docs, error messages and style

review wip

adding an all null column should not break anything; bugfix

review wip

update test logic

SparkQA commented:
**[Test build #88494 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88494/testReport)** for PR 20829 at commit [`e7e26f0`](https://github.com/apache/spark/commit/e7e26f0ccc149739642aa0a7c52ba473a0307b75).

Commit:
No need to compare with anything; just call assemble()

SparkQA commented:
**[Test build #88495 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88495/testReport)** for PR 20829 at commit [`4c99003`](https://github.com/apache/spark/commit/4c99003e05cdd9538ae44b392c652a2863f4cb71).

SparkQA commented:
**[Test build #88494 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88494/testReport)** for PR 20829 at commit [`e7e26f0`](https://github.com/apache/spark/commit/e7e26f0ccc149739642aa0a7c52ba473a0307b75).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88494/
Test PASSed.

SparkQA commented:
**[Test build #88495 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88495/testReport)** for PR 20829 at commit [`4c99003`](https://github.com/apache/spark/commit/4c99003e05cdd9538ae44b392c652a2863f4cb71).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88495/
Test PASSed.

Commit:
handle NaNs too; update tests; style

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/feature/VectorAssemblerSuite.scala:
```diff
@@ -18,56 +18,68 @@
-    val df = Seq(
-      (0, 0.0, Vectors.dense(1.0, 2.0), "a", Vectors.sparse(2, Array(1), Array(3.0)), 10L)
-    ).toDF("id", "x", "y", "name", "z", "n")
+    val df = dfWithNulls.filter("id1 == 1").withColumn("id", col("id1"))
```
nit: If this is for consolidation, I'm actually against this little change since it obscures what this test is doing and moves the input Row farther from the expected output row.

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/feature/VectorAssemblerSuite.scala:
```diff
@@ -147,4 +159,88 @@ class VectorAssemblerSuite
+      .setInputCols(Array("x", "y", "z", "n"))
+      .setOutputCol("features")
+
+    def run_with_metadata(mode: String, additional_filter: String = "true"): Dataset[_] = {
```
style: use camelCase

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/feature/VectorAssemblerSuite.scala:
```diff
@@ -147,4 +159,88 @@ class VectorAssemblerSuite
+
+    def run_with_metadata(mode: String, additional_filter: String = "true"): Dataset[_] = {
+      val attributeY = new AttributeGroup("y", 2)
+      val subAttributesOfZ = Array(NumericAttribute.defaultAttr, NumericAttribute.defaultAttr)
```
unused

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/feature/VectorAssemblerSuite.scala:
```diff
@@ -147,4 +159,88 @@ class VectorAssemblerSuite
+      output.collect()
+      output
+    }
+    def run_with_first_row(mode: String): Dataset[_] = {
```
style: Put empty line between functions

jkbradley reviewed (commented):
Thanks for the updates!  I mostly have style comments at this point.

yogeshg commented on mllib/src/main/scala/org/apache/spark/ml/feature/VectorAssembler.scala:
```diff
@@ -49,32 +55,64 @@ class VectorAssembler @Since("1.4.0") (@Since("1.4.0") override val uid: String)
+    }
+    val vectorColsLengths = VectorAssembler.getLengths(dataset, vectorCols, $(handleInvalid))
+
+    val featureAttributesMap = $(inputCols).toSeq.map { c =>
```
We need the map to find out the length of vectors, unless there's a way to do this in one mapping way, I think it might be better than to call first a `map` and then a `flatMap`.

Commit:
hint in comments

SparkQA commented:
**[Test build #88829 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88829/testReport)** for PR 20829 at commit [`081b5c0`](https://github.com/apache/spark/commit/081b5c03d5364e359b1eceb0fc289d85e940f1c9).

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/feature/VectorAssembler.scala:
```diff
@@ -136,34 +181,106 @@ class VectorAssembler @Since("1.4.0") (@Since("1.4.0") override val uid: String)
+   * @return map of column names to lengths
+   */
+  private[feature] def getVectorLengthsFromFirstRow(
+      dataset: Dataset[_], columns: Seq[String]): Map[String, Int] = {
```
This still isn't correct; look at other MLlib code for examples.  It should be:
```
private[feature] def getVectorLengthsFromFirstRow(
    dataset: Dataset[_],
    columns: Seq[String]): Map[String, Int] = {
```

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/feature/VectorAssembler.scala:
```diff
@@ -136,34 +181,106 @@ class VectorAssembler @Since("1.4.0") (@Since("1.4.0") override val uid: String)
+    }
+  }
+
+  private[feature] def getLengths(dataset: Dataset[_], columns: Seq[String],
```
Ditto

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/feature/VectorAssembler.scala:
```diff
@@ -136,34 +181,106 @@ class VectorAssembler @Since("1.4.0") (@Since("1.4.0") override val uid: String)
+      case e: NullPointerException => throw new NullPointerException(
+        s"""Encountered null value while inferring lengths from the first row. Consider using
+           |VectorSizeHint to add metadata for columns: ${columns.mkString("[", ", ", "]")}.
+           |""".stripMargin.replaceAll("\n", " ") + e.toString)
```
ping

yogeshg replied:
"\n" should be replaced with " " and then a message. Unless I misunderstand something.

jkbradley replied:
Oh, I see; you're right.  Does this mean the latest change removes the space though?

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/feature/VectorAssembler.scala:
```diff
@@ -136,34 +181,106 @@ class VectorAssembler @Since("1.4.0") (@Since("1.4.0") override val uid: String)
+      case e: NoSuchElementException => throw new NoSuchElementException(
+        s"""Encountered empty dataframe while inferring lengths from the first row. Consider using
+           |VectorSizeHint to add metadata for columns: ${columns.mkString("[", ", ", "]")}.
+           |""".stripMargin.replaceAll("\n", " ") + e.toString)
```
ping

jkbradley reviewed (commented):
The changes look good, but there are a few unaddressed comments.

SparkQA commented:
**[Test build #88829 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88829/testReport)** for PR 20829 at commit [`081b5c0`](https://github.com/apache/spark/commit/081b5c03d5364e359b1eceb0fc289d85e940f1c9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88829/
Test PASSed.

Commit:
remaining style changes in review

Commit:
restore old example

SparkQA commented:
**[Test build #88835 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88835/testReport)** for PR 20829 at commit [`bf277be`](https://github.com/apache/spark/commit/bf277be99b6574ff0bd57bf4af2e2f6719a85ccf).

SparkQA commented:
**[Test build #88834 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88834/testReport)** for PR 20829 at commit [`134bd1e`](https://github.com/apache/spark/commit/134bd1e21dd7ff3f1c0577fad9a6516847e59f3c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88834/
Test PASSed.

SparkQA commented:
**[Test build #88835 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88835/testReport)** for PR 20829 at commit [`bf277be`](https://github.com/apache/spark/commit/bf277be99b6574ff0bd57bf4af2e2f6719a85ccf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88835/
Test PASSed.

jkbradley commented:
LGTM
Merging with master
Thanks @yogeshg for the PR and @WeichenXu123 for taking a look!
